lib/rex/post/meterpreter/channel.rb

```diff
@@ -153,11 +153,16 @@ def initialize(client, cid, type, flags, packet, **_)
   def self.finalize(client, cid)
     proc {
       unless cid.nil?
-        begin
-          self._close(client, cid)
-        rescue => e
-          elog("finalize method for Channel failed", error: e)
+        deferred_close_proc = proc do
+          begin
+            self._close(client, cid)
+          rescue => e
+            elog("finalize method for Channel failed", error: e)
+          end
         end
+
+        # Schedule the finalizing logic out-of-band; as this logic might be called in the context of a Signal.trap, which can't synchronize mutexes
+        client.framework.sessions.schedule(deferred_close_proc)
```

Comment:
This is fine to call, as the `schedule` implementation only does a push to a queue:

https://github.com/rapid7/metasploit-framework/blob/dbda20d40b44d4557d6f736b1056ae4d4d83c858/lib/msf/core/session_manager.rb#L170-L176

The Ruby queue is thread safe, and can be called safely in a trap context

https://github.com/rapid7/metasploit-framework/blob/dbda20d40b44d4557d6f736b1056ae4d4d83c858/lib/msf/core/session_manager.rb#L30

From the docs:
> Thread::Queue#push and Thread::SizedQueue#push (since Ruby 2.1)
> https://github.com/ruby/ruby/blob/012faccf040344360801b0fa77e85f9c8a3a4b2c/doc/signals.rdoc#label-Commonly+safe+operations+inside+Signal.trap+blocks

Which is the case here:

``
>> framework.sessions.send(:scheduler_queue).class
=> Thread::Queue
```